java-frontend/src/main/java/org/sonar/java/se/checks/DivisionByZeroCheck.java

```diff
@@ -0,0 +1,300 @@
+      List<SymbolicValue> symbolicValues = programState.peekValues(2);
+      SymbolicValue var = symbolicValues.get(0);
+      SymbolicValue expr = symbolicValues.get(1);
+      checkExpression(tree, var, expr);
+    }
+
+    @Override
+    public void visitBinaryExpression(BinaryExpressionTree tree) {
+      List<SymbolicValue> symbolicValues;
+      switch (tree.kind()) {
+        case MULTIPLY:
+        case PLUS:
+        case MINUS:
+        case DIVIDE:
+        case REMAINDER:
+          symbolicValues = programState.peekValues(2);
+          checkExpression(tree, symbolicValues.get(1), symbolicValues.get(0));
+          break;
+        case GREATER_THAN:
+        case GREATER_THAN_OR_EQUAL_TO:
+        case LESS_THAN:
+        case LESS_THAN_OR_EQUAL_TO:
+          symbolicValues = programState.peekValues(2);
+          setAsUndetermined(symbolicValues.get(1), tree.leftOperand());
+          setAsUndetermined(symbolicValues.get(0), tree.rightOperand());
+          break;
+        default:
+          // do nothing
+      }
+    }
+
+    private void setAsUndetermined(SymbolicValue sv, Tree tree) {
+      programState = programState.addConstraint(sv, new ZeroConstraint(tree, Status.UNDETERMINED));
+    }
+
+    private void checkExpression(Tree tree, SymbolicValue leftOp, SymbolicValue rightOp) {
+      switch (tree.kind()) {
+        case MULTIPLY:
+        case MULTIPLY_ASSIGNMENT:
+          handleMultiply(leftOp, rightOp);
+          break;
+        case PLUS:
+        case PLUS_ASSIGNMENT:
+        case MINUS:
+        case MINUS_ASSIGNMENT:
+          handlePlusMinus(leftOp, rightOp);
+          break;
+        case DIVIDE:
+        case DIVIDE_ASSIGNMENT:
+        case REMAINDER:
+        case REMAINDER_ASSIGNMENT:
+          if (isZero(rightOp)) {
+            reportIssue(tree);
+          } else {
+            handleDivide(tree.kind(), leftOp, rightOp);
+          }
+          break;
+        default:
+          // can not be reached
+      }
+    }
+
+    private boolean isZero(SymbolicValue symbolicValue) {
+      return hasStatus(symbolicValue, Status.ZERO);
+    }
+
+    private boolean isNonZero(SymbolicValue symbolicValue) {
+      return hasStatus(symbolicValue, Status.NON_ZERO);
+    }
+
+    private boolean hasStatus(SymbolicValue symbolicValue, Status status) {
+      return programState.getConstraintWithStatus(symbolicValue, status) != null;
+    }
+
+    private void handleMultiply(SymbolicValue left, SymbolicValue right) {
+      if (isZero(left) || isZero(right)) {
+        deferConstraint(Status.ZERO);
+      } else if (isNonZero(left) && isNonZero(right)) {
+        deferConstraint(Status.NON_ZERO);
+      }
+    }
+
+    private void handlePlusMinus(SymbolicValue left, SymbolicValue right) {
+      if (isZero(left) && isZero(right)) {
+        deferConstraint(Status.ZERO);
+      } else if ((isZero(left) && isNonZero(right)) || (isNonZero(left) && isZero(right))) {
```

Comment:
This is in essence an assignement. As such this method should be simplified to handle the fact that if left is zero, then resulting SV is the right SV.